projects/packages/jetpack-mu-wpcom/src/features/block-patterns/class-wpcom-block-patterns-utils.php

```diff
@@ -0,0 +1,149 @@
+	 *
+	 * @param string $patterns_slug A slug for a patterns source site, e.g., `block_patterns`.
+	 * @return string locale slug
+	 */
+	public function get_patterns_cache_key( $patterns_slug ) {
+		return sha1(
+			implode(
+				'_',
+				array(
+					$patterns_slug,
+					A8C_ETK_PLUGIN_VERSION,
+					$this->get_block_patterns_locale(),
+				)
+			)
+		);
+	}
+
+	/**
+	 * Get the locale to be used for fetching block patterns
+	 *
+	 * @return string locale slug
+	 */
+	public function get_block_patterns_locale() {
+		// Block patterns display in the user locale.
+		$language = get_user_locale();
+		return \A8C\FSE\Common\get_iso_639_locale( $language );
+	}
+
+	/**
+	 * Check for block type values in the pattern_meta tag.
+	 * When tags have a prefix of `block_type_`, we expect the remaining suffix to be a blockType value.
+	 * We'll add these values to the `(array) blockType` options property when registering the pattern
+	 * via `register_block_pattern`.
+	 *
+	 * @param array $pattern A pattern with a 'pattern_meta' array.
+	 *
+	 * @return array         An array of block types defined in pattern meta.
+	 */
+	public function maybe_get_pattern_block_types_from_pattern_meta( $pattern ) {
+		$block_types = array();
+
+		if ( ! isset( $pattern['pattern_meta'] ) || empty( $pattern['pattern_meta'] ) ) {
+			return $block_types;
+		}
+
+		foreach ( $pattern['pattern_meta'] as $pattern_meta => $value ) {
+			// Match against tags starting with `block_type_`.
+			$split_slug = preg_split( '/^block_type_/', $pattern_meta );
+
+			if ( isset( $split_slug[1] ) ) {
+				$block_types[] = $split_slug[1];
+			}
+		}
+
+		return $block_types;
+	}
+
+	/**
+	 * Return pattern post types based on the pattern's blockTypes.
+	 *
+	 * @param array $pattern A pattern array such as is passed to `register_block_pattern`.
+	 *
+	 * @return array $postTypes An array of post type names such as is passed to `register_block_pattern`.
+	 */
+	public function get_pattern_post_types_from_pattern( $pattern ) {
+		$post_types = array_key_exists( 'postTypes', $pattern ) ? $pattern['postTypes'] : array();
+		if ( $post_types ) {
+			// If some postTypes are explicitly set then respect the pattern author's intent.
+			return $post_types;
+		}
+
+		$block_types       = array_key_exists( 'blockTypes', $pattern ) ? $pattern['blockTypes'] : array();
+		$block_types_count = count( $block_types );
+		// If all of a patterns blockTypes are template-parts then limit the postTypes to just
+		// the template related types and to pages - this is to avoid the pattern appearing in
+		// the inserter for posts and other post types. Pages are included because it's not unusual
+		// to use a blank template and add a specific header and footer to a page.
+		if ( $block_types_count && count(
+			array_filter(
+				$block_types,
+				function ( $block_type ) {
+					return preg_match( '#core/template-part/#', $block_type );
+				}
+			)
+		) === $block_types_count ) {
```

Comment:
This is really hard to read. Could we pull out the `count( array_filter() )` and save that in a variable and compare that?